### ISSUE TITLE
Firefox Video tiles freeze after turning on/off

### ISSUE THREAD

**Issue by edwardbc** [aws/amazon-chime-sdk-js#347]
**Meeting and attendee IDs:**
Meeting ID: 608e301f-7825-4977-8e40-af53b1a5655

**Timeline**
1. Chrome user joins the meeting and starts Video
2. Chrome user starts Video
3. Firefox user joins the meeting
4. Firefox user starts video (Warning 1)
5. Firefox user stops video, 
6. Firefox user starts video again (Warning 2)
7. Remote Video tile on Firefox freezes

*Warning 1:* ```2020-05-06T21:12:22.445Z [WARN] SDK - removing existing tiles 1 with same attendee id e3a8fe33-eec2-4b92-8bb9-b14a2a237bfc```

*Warning 2:* ```2020-05-06T21:13:13.897Z [WARN] SDK - removing existing tiles 3 with same attendee id e3a8fe33-eec2-4b92-8bb9-b14a2a237bfc```

**Describe the bug**
On Firefox, when joining a meeting with an existing video tile, then turning on video a warning is displayed (first sign), then when the other user turns off and on their video it freezes and `videoTileWasRemoved` never gets called. Subsequent attempts to display the remote video tile fail and it's never restored on Firefox

NOTE: Doesn't always happen when Firefox joins the meeting first. There needs to be an active video tile in place when Firefox joins to replicate this easily.
Additionally, on a separate app, calling `.audioVideo.getAllVideoTiles()` on the console after the Firefox user joins with Video, before toggling any tiles on the other side, it returns three tiles instead of two on Firefox, the third having a `tileState` full of `null` values. I couldn't replicate this specific case on the Browser demo, but might be related.

**Logs**
From browser demo. Initially replicated in a different app and flow.
[ChromeUser.log](https://github.com/aws/amazon-chime-sdk-js/files/4589558/ChromeUser.log)
[FirefoxUser.log](https://github.com/aws/amazon-chime-sdk-js/files/4589559/FirefoxUser.log)

**Platform**
 - OS: 
 - Browser:  Firefox
 - Browser Version:  76.0
 - Amazon Chime SDK GitHub Commit: 1.4.0

**Comment by edwardbc** [comment]
There's a third warning that is sometimes displayed when enabling video again on the non-Firefox attendee
```2020-05-06T21:49:30.851Z [WARN] ChimeMeetingLogs - no stream found for tile=null```

**Comment by lhristov** [comment]
I have the same issue. Any idea when this will be resolved?
Thanks!

**Comment by ltrung** [comment]
@edwardbc Thank you for reporting this. However, I could not reproduce this issue in our browser demo app. Here are the steps I tried:
- Chrome user joins the meeting
- Chrome user starts Video
- Firefox user joins the meeting
- Firefox user starts video (I do not see any warning here)
- Firefox user stops video
- Firefox user starts video again (No warning either)
- The remote video tile from Chrome on Firefox is not frozen here.

Browser version:
- OS: Mac
- Chrome: 81
- Firefox: 68
- Amazon Chime SDK GitHub Commit: 1.5.0

**Comment by edwardbc** [comment]
@ltrung can you try also turning on and off the camera on both sides once both users have enabled video?

I can easily replicate on Firefox *76.0* the browser demo, updated to commit:
`dbbac6600926371620f0696d2dc69a22c060e9c1`

We keep getting errors like these:
```[WARN] SDK - removing existing tiles 1 with same attendee id 455dd0a9-a015-4810-9c80-814cffdfcf75```

Seems Firefox is getting duplicate streams for some strange reason.

Separately, in our app we've  alsonoticed that by just starting the video tiles on Firefox alone in a meeting, there's additional video tracks passed to `CreatePeerConnectionTask.prototype.addRemoteVideoTrack(track, stream)` coming from `trackAddedHandler` on multiple `track` events. This causes errors as it seems `streams` comes undefined due this track duplication.

**Comment by alochbau-amzn** [comment]
I was able to reproduce this on Firefox 76.0, but not on Firefox 75.0b9.  It looks like there was a change in the webrtc library in Firefox 76.0 which causes it to behave differently.  Until we get this sorted out, can you try switching to Firefox 75 and use that version?

Thanks

**Comment by Ardenexal** [comment]
We are also getting this issue seems to be caused by a change to Firefox in 76.0, telling users to downgrade to an older version of firefox is not really an option for us. hoping there will be an update on this soon

**Comment by rebecca-networkapp** [comment]
In Firefox 76.0 we had the issue that the streams of other participants vanished when I started my own camera. 

We solved this by applying the following hack in `src/task/CreatePeerConnectionTask.ts`

On top of the file add (so TypeScript won't complain about InstallTrigger, not being defined):
```
declare const InstallTrigger: any;
```
Replace line `184-192` in `src/task/CreatePeerConnectionTask.ts` with the following:
```
var isFirefox = typeof InstallTrigger !== 'undefined';
if(!isFirefox){
    // TODO: in case a previous tile with the same attendee id hasn't been cleaned up
    // we do it here to avoid showing a duplicate tile. We should try to understand
    // why this can happen, so adding a log statement to track this and learn more.
    const tilesRemoved = this.context.videoTileController.removeVideoTilesByAttendeeId(attendeeId);
    if (tilesRemoved.length > 0) {
      this.logger.warn(
        `removing existing tiles ${tilesRemoved} with same attendee id ${attendeeId}`
      );
    }
}
else
{
    for (const tile of this.context.videoTileController.getAllVideoTiles()) {
      const state = tile.state();
      if (state.boundAttendeeId === attendeeId) {
        return;
      }
    }
}
```

Hoop this helps somebody.

**Comment by edwardbc** [comment]
Seems the fixes from pull https://github.com/aws/amazon-chime-sdk-js/pull/389 fix the issue with the additional track on FF 76 and tiles are restored correctly so far. Thanks!!!

On the other side, there's an error that it's triggered now every now and then after testing the above scenario and turning on and off the video on both clients , eventually the following error can be seen: 

Log sample 
```
INFO: videoAvailabilityDidChange  : { remoteVideoAvailable: false, canStartLocalVideo: true }

ERROR: TypeError: this.removeTrackRemovedEventListeners[track.id] is not a function

INFO: videoAvailabilityDidChange  ()  { remoteVideoAvailable: true, canStartLocalVideo: true }
```
On:
https://github.com/aws/amazon-chime-sdk-js/blob/5a9d8d0a6b6c0809c7a856a1e3da89b1706dfdb0/src/task/CreatePeerConnectionTask.ts#L263

**Comment by alochbau-amzn** [comment]
@edwardbc glad to hear the original issue is fixed.

For this new issue can you provide some more information?
Did you still only have the two clients, one Firefox and one Chrome?
toggleVideo is for the local video or for the remote video, and what exactly is it doing?

Debug level logs would be helpful too.

Thanks

**Comment by edwardbc** [comment]
**Timeline**

1. Chrome user joins the meeting and starts Video
2. Firefox user joins the meeting and starts Video
3. Chrome user turns on, off, on, off and so on.
4. Error is thrown on Firefox console. Video tile might freeze sometimes.

To start and stop video we call `startLocalVideoTile()` and `stopLocalVideoTile()` 

**Logs**
I'm attaching both Chrome and Firefox debug logs. On Firefox you can see the error at the end of the file. Around timestamp `2020-05-21T17:25:46.718Z`


[ChromeUser-debug.log](https://github.com/aws/amazon-chime-sdk-js/files/4664022/ChromeUser-debug.log)
[FirefoxUser-debug.log](https://github.com/aws/amazon-chime-sdk-js/files/4664024/FirefoxUser-debug.log)

**Comment by alochbau-amzn** [comment]
This looks like a different issue.

I am not positive on what exactly the problem is, but there are a couple things I see in the Firefox logs.  It looks like you are trying to start local video before you received the videoAvailibilityDidChange event which says it is ok to start it.

```
2020-05-21T17:25:40.911Z [INFO] ChimeMeetingLogs - API/DefaultAudioVideoFacade/69f16d9e-3260-4ae7-9e1d-066c014ca1ae/6cbd4450-4c85-479b-bb2e-6d6756855700/startLocalVideoTile null -> 1 amazon-chime-sdk.js:1185:26
```
```
2020-05-21T17:25:41.649Z [DEBUG] ChimeMeetingLogs - [DefaultStatsCollector] meeting_ping_pong_clock_skew: {"call_id":"69f16d9e-3260-4ae7-9e1d-066c014ca1ae","client_type":"amazon-chime-sdk-js","metric_type":"latency"} amazon-chime-sdk.js:1182:26
=> videoAvailabilityDidChange 
Object { remoteVideoAvailable: true, canStartLocalVideo: true }
modules.js:7699:17
```
That could be tripping up the tileController since I see some other tile/track warnings like this one on the first time far video was stopped.  It seems that the tile is no longer there or associated with the video track.
```
2020-05-21T17:25:45.443Z [INFO] ChimeMeetingLogs - video track ended, removing tile=null id={d70b4bf6-7631-e045-bc37-f8c403d2c33d} stream=null amazon-chime-sdk.js:1185:26
2020-05-21T17:25:45.443Z [WARN] ChimeMeetingLogs - no stream found for tile=null amazon-chime-sdk.js:1179:26
```

Could you try modifying your code to wait before calling startLocalVideoTile?

I tried to reproduce this with the unmodified browser demo and was able to  start/stop many times on the Chrome side without any errors.

**Comment by montamal** [comment]
I don't have specifics on the timeline or logs to prove it, but I'm simply here to +1 on the issues with the latest release of Firefox.

Users on FF76, using v1.6.2 of the SDK are seeing duplicate tiles and getting blank(black) or frozen tiles whenever other users join/leave a meeting. Will try and collect some logs on it.

Thanks @alochbau-amzn for your support.

**Comment by alochbau-amzn** [comment]
@montamal  Would be best if you try to collect logs while running the latest from master.  A fix for FF76 missing/grey tiles is in master but is not in a release yet.  I am hoping those changes fix the issues you are seeing.

Thanks

**Comment by edwardbc** [comment]
@alochbau-amzn I've created a sample vanilla app, based on the documentation code where both Firefox issues can be easily replicated. The second scenario where FF duplicates tiles doesn't seem to be FF76 only.

Hope it helps

**Comment by alochbau-amzn** [comment]
@edwardbc Thanks, this will be very helpful!

Having multiple browsers open on my Mac all trying to use WebRTC doesn't seem to work very good.  Seems they are fighting for the camera.  I was able to recreate Scenario 1 with it, though when I joined with Firefox local video didn't initially start so the recreation steps were slightly different.

 When trying to run scenario 2 on my one system, the Chrome local video stopped when Firefox joined the meeting and I never saw duplicate tiles.  I am going to play around with it some and try with 2 computers to see if that makes it easier to recreate.
